Overhaul the `BASICS` part of the documentation (#50932)
LLNL/spack: lib/spack/docs/spec_syntax.rst

Here is a simplified dependency graph for ``mpileaks``:

.. graphviz::

   digraph {
       node[
         fontname=Monaco,
         penwidth=2,
         fontsize=124,
         margin=.4,
         shape=box,
         fillcolor=lightblue,
         style="rounded,filled"
       ]

       mpileaks -> { mpich callpath }
       callpath -> { mpich dyninst }
       dyninst  -> libdwarf -> libelf
       dyninst  -> libelf
   }

Each box above is a package, and each arrow represents a dependency on some other package.
For example, we say that the package ``mpileaks`` *depends on* ``callpath`` and ``mpich``.
``mpileaks`` also depends *indirectly* on ``dyninst``, ``libdwarf``, and ``libelf``, in that these libraries are dependencies of ``callpath``.
To install ``mpileaks``, Spack has to build all of these packages.
Dependency graphs in Spack have to be acyclic, and the *depends on* relationship is directional, so this is a *directed, acyclic graph* or *DAG*.

The package name identifier in the spec is the root of some dependency DAG, and the DAG itself is implicit.
Spack knows the precise dependencies among packages, but users do not need to know the full DAG structure.

> Each ``^`` in the full spec refers to some *transitive* dependency of the root package.

Reviewer: ```suggestion
Each ``^`` in the full spec refers to a *transitive* dependency of the root package.
```

> Each ``%`` refers to some *direct* dependency, either of the root, or of the last defined transitive dependency .

Reviewer: ```suggestion
Each ``%`` refers to a *direct* dependency, either of the root, or of the last defined transitive dependency .
```

> Spack takes care of allowing only a single configuration of each package, where that is needed for consistency.

Reviewer: "takes care of allowing" sounds weird

Above, both ``mpileaks`` and ``callpath`` depend on ``mpich``, but ``mpich`` appears only once in the DAG.
You cannot build an ``mpileaks`` version that depends on one version of ``mpich`` *and* on a ``callpath`` version that depends on some *other* version of ``mpich``.
In general, such a configuration would likely behave unexpectedly at runtime, and Spack enforces this to ensure a consistent runtime environment.

> The point of specs is to abstract this full DAG from Spack users.

Reviewer: ```suggestion
The purpose of specs is to abstract this full DAG away from Spack users.
```

> A user that does not care about the DAG at all, can refer to mpileaks by simply writing:

Reviewer: ```suggestion
A user who does not care about the DAG at all, can refer to ``mpileaks`` by simply writing:
```

``0.19``, then the spec will be shown as ``@git.abcdefg=0.19``.

If the git ref is not exactly a tag, then the distance to the nearest tag
is also part of the resolved version. ``@git.abcdefg=0.19.git.8`` means
that the commit is 8 commits away from the ``0.19`` tag.

In cases where Spack cannot resolve a sensible version from a git ref,
users can specify the Spack version to use for the git ref. This is done
by appending ``=`` and the Spack version to the git ref. For example:

.. code-block:: sh

   foo@git.my_ref=3.2 # use the my_ref tag or branch, but treat it as version 3.2 for version comparisons
   foo@git.abcdef1234abcdef1234abcdef1234abcdef1234=develop # use the given commit, but treat it as develop for version comparisons

Details about how versions are compared and how Spack determines if
one version is less than another are discussed in the developer guide.

.. _basic-variants:

--------
Variants
--------

Variants are named options associated with a particular package and are typically used to enable or disable certain features at build time.
They are optional, as each package must provide default values for each variant it makes available.

> The names of variants available for a particular package depend on what was provided by the package author.

Reviewer: ```suggestion
The variants available for a particular package are defined by the package author.
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Spack allows variants to propagate their value to the package's
dependencies by using ``++``, ``--``, and ``~~`` for boolean variants.
For example, for a ``debug`` variant:

.. code-block:: sh

    mpileaks ++debug   # enabled debug will be propagated to dependencies
    mpileaks +debug    # only mpileaks will have debug enabled

To propagate the value of non-boolean variants Spack uses ``name==value``.
For example, for the ``stackstart`` variant:

.. code-block:: sh

    mpileaks stackstart==4   # variant will be propagated to dependencies
    mpileaks stackstart=4    # only mpileaks will have this variant value

Spack also allows variants to be propagated from a package that does
not have that variant.

^^^^^^^^^^^^^^^^^
Binary Provenance
^^^^^^^^^^^^^^^^^

Spack versions are paired to attributes that determine the source code Spack
will use to build. Checksummed assets are preferred but there are a few
notable exceptions such as git branches and tags i.e ``pkg@develop``.
These versions do not naturally have source provenance because they refer to a range
of commits (branches) or can be changed outside the spack packaging infrastructure

> (tags). Without source provenace we can not have binary provenance.

Reviewer: ```suggestion
(tags). Without source provenance we cannot have binary provenance.
```

   be necessary.

   Above all else, the most robust way to ensure binaries have their desired commits is to provide
   the SHAs via user-specs or config i.e. ``commit=<SHA>``.

--------------
Compiler Flags
--------------

Compiler flags are specified using the same syntax as non-boolean variants,
but fulfill a different purpose. While the function of a variant is set by
the package, compiler flags are used by the compiler wrappers to inject
flags into the compile line of the build. Additionally, compiler flags can
be inherited by dependencies by using ``==``.
``spack install libdwarf cppflags=="-g"`` will install both libdwarf and
libelf with the ``-g`` flag injected into their compile line.

Notice that the value of the compiler flags must be quoted if it
contains any spaces. Any of ``cppflags=-O3``, ``cppflags="-O3"``,
``cppflags='-O3'``, and ``cppflags="-O3 -fPIC"`` are acceptable, but
``cppflags=-O3 -fPIC`` is not. Additionally, if the value of the
compiler flags is not the last thing on the line, it must be followed
by a space. The command ``spack install libelf cppflags="-O3"%intel``
will be interpreted as an attempt to set ``cppflags="-O3%intel"``.

> The six compiler flags are injected in the order of implicit make commands

Reviewer: ```suggestion
The six compiler flags are injected in the same order as implicit make commands
```

A complete list of the microarchitectures known to Spack can be obtained in the following way:

.. command-output:: spack arch --known-targets

When a spec is installed, Spack matches the compiler being used with the microarchitecture being targeted to inject appropriate optimization flags at compile time.
Giving a command such as the following:

.. code-block:: console

   $ spack install zlib%gcc@14.2.0 target=icelake

will produce compilation lines similar to:

.. code-block:: console

   $ /usr/bin/gcc-14 -march=icelake-client -mtune=icelake-client -c ztest10532.c
   $ /usr/bin/gcc-14 -march=icelake-client -mtune=icelake-client -c -fPIC -O2 ztest10532.
   ...

where the flags ``-march=icelake-client -mtune=icelake-client`` are injected by Spack based on the requested target and compiler.

If Spack knows that the requested compiler can't optimize for the current target
or can't build binaries for that target at all, it will exit with a meaningful error message:

.. code-block:: console

   $ spack install zlib%gcc@5.5.0 target=icelake
   ==> Error: cannot produce optimized binary for micro-architecture "icelake" with gcc@5.5.0 [supported compiler versions are 8:]

> When instead an old compiler is selected on a recent enough microarchitecture but there is
> no explicit ``target`` specification, Spack will optimize for the best match it can find instead

Reviewer: ```suggestion
Conversely, if an old compiler is selected for a newer microarchitecture, Spack will optimize for the best match it can find instead
```

   SPDX-License-Identifier: (Apache-2.0 OR MIT)

.. _sec-specs:

===========
Spec Syntax
===========

Spack has a specific syntax to describe package constraints.
Each constraint is individually referred to as a *spec*. Spack uses specs to:

1. Refer to a particular build configuration of a package, or
2. Express requirements, or preferences, on packages via configuration files, or
3. Query installed packages, or buildcaches

Specs are more than a package name and a version; you can use them to
specify the compiler, compiler version, architecture, compile options,
and dependency options for a build. In this section, we'll go over
the full syntax of specs.

Here is an example of a moderately complex spec:

.. code-block:: none

   mpileaks @1.2:1.4 +debug ~qt target=x86_64 %gcc@4.7.5 ^callpath @1.1 %gcc@4.7.2

If used to install a package, this will install the ``mpileaks``
library at some version between ``1.2`` and ``1.4`` (inclusive),
built using ``gcc`` at version 4.7.5 for a generic ``x86_64`` architecture,
with debug options enabled, and without Qt support. Additionally, it
says to link it with the ``callpath`` library (which it depends on),

> and to build callpath with ``gcc`` 4.7.2. Most specs will not be as

Reviewer: The spec does not say "to link it with", it says it has to depend on it.

Maybe also update the compiler example to something more sensible, like `%gcc@13`.

complicated as this one, but this is a good example of what is
possible with specs.

More formally, a spec consists of the following pieces:

* Package name identifier (``mpileaks`` above)
* ``@`` Optional version specifier (``@1.2:1.4``)
* ``+`` or ``-`` or ``~`` Optional variant specifiers (``+debug``,
  ``-qt``, or ``~qt``) for boolean variants. Use ``++`` or ``--`` or
  ``~~`` to propagate variants through the dependencies (``++debug``,
  ``--qt``, or ``~~qt``).
* ``name=<value>`` Optional variant specifiers that are not restricted to
  boolean variants. Use ``name==<value>`` to propagate variant through the
  dependencies.
* ``name=<value>`` Optional compiler flag specifiers. Valid flag names are
  ``cflags``, ``cxxflags``, ``fflags``, ``cppflags``, ``ldflags``, and ``ldlibs``.
  Use ``name==<value>`` to propagate compiler flags through the dependencies.
* ``target=<value> os=<value>`` Optional architecture specifier

>   (e.g., ``target=haswell os=CNL10``)

Reviewer: platform is missing, os is an unfamiliar example to all readers.

* ``%`` Direct dependency specs. Specs the user knows are not merely present in
  the graph, but depended on directly by the previous node.
* ``^`` Dependency specs (e.g., ``^callpath@1.1``). These dependencies may appear
  anywhere in the link/run dependencies of the root, or in the direct build
  dependencies.

There are two things to notice here. The first is that specs are
recursively defined. That is, each dependency after ``%`` or ``^`` is
a spec itself. The second is that everything is optional *except* for

> the initial package name identifier. Users can be as vague or as

Reviewer: not the case, depends on context...

   digraph {
       mpileaks -> mpich
       mpileaks -> callpath -> mpich
       callpath -> dyninst
       dyninst  -> libdwarf -> libelf
       dyninst  -> libelf
   }

Each box above is a package, and each arrow represents a dependency on
some other package. For example, we say that the package ``mpileaks``
*depends on* ``callpath`` and ``mpich``. ``mpileaks`` also depends
*indirectly* on ``dyninst``, ``libdwarf``, and ``libelf``, in that
these libraries are dependencies of ``callpath``. To install
``mpileaks``, Spack has to build all of these packages. Dependency
graphs in Spack have to be acyclic, and the *depends on* relationship
is directional, so this is a *directed, acyclic graph* or *DAG*.

The package name identifier in the spec is the root of some dependency
DAG, and the DAG itself is implicit. Spack knows the precise
dependencies among packages, but users do not need to know the full
DAG structure. Each ``^`` in the full spec refers to some dependency
of the root package. Spack will raise an error if you supply a name
after ``^`` that the root does not actually depend on (e.g., ``mpileaks
^emacs@23.3``). Each ``%`` refers to some direct dependency, and Spack will

> similarly raise an error if that relationship is invalid.

Reviewer: why so much focus on "raising an error" :thinking: it's about the meaning of ^ and %?

Spack further simplifies things by only allowing one configuration of

> each package within the link/run + direct build dependencies of a

Reviewer: dependency types were not introduced. If the intention is to be vague, just talk about runtime dependencies.

Author: So far, I just moved the content up for `Spec Syntax`. I was planning to improve on it in a following PR, and add diagrams etc.

Reviewer: I would just make all changes in one PR. Having URLs change constantly is bad for search engines and sharing links to docs/sections.

Version ranges are inclusive, so this example includes both ``1.0``
and any ``1.5.x`` version. Version ranges can be unbounded, e.g., ``@:3``
means any version up to and including ``3``. This would include ``3.4``
and ``3.4.2``. Similarly, ``@4.2:`` means any version above and including
``4.2``. As a shorthand, ``@3`` is equivalent to the range ``@3:3`` and
includes any version with major version ``3``.

Versions are ordered lexicographically by their components. For more details
on the order, see :ref:`the packaging guide <version-comparison>`.

Notice that you can distinguish between the specific version ``@=3.2`` and
the range ``@3.2``. This is useful for packages that follow a versioning
scheme that omits the zero patch version number: ``3.2``, ``3.2.1``,
``3.2.2``, etc. In general, it is preferable to use the range syntax
``@3.2``, because ranges also match versions with one-off suffixes, such as
``3.2-custom``.

A version specifier can also be a list of ranges and specific versions,
separated by commas. For example, ``@1.0:1.5,=1.7.1`` matches any version
in the range ``1.0:1.5`` and the specific version ``1.7.1``.

-----------------

> Binary Provenance

Reviewer: Feels out of place. It's highly specific, and the generic variants section is below it.

Author: Moved down in the `Variant` section
